File: 101-109-smallproblems/medium1/stackmachine.rb

def minilang(str)
  stack = []
  register = 0

  str.split(' ').each do |command|
    begin
      case command
      when /\d+/
        register = command.to_i
      when 'PUSH'
        stack.push register
      when 'POP'
        register = stack.pop
      when 'ADD'
        register += stack.pop
      when 'SUB'
        register -= stack.pop
      when 'MULT'
        register *= stack.pop
      when 'DIV'
        register /= stack.pop
      when 'MOD'
        register = register % stack.pop
      when 'PRINT'
        puts register
      else
        raise ArgumentError.new("'#{command}' is an invalid command.")
      end
    rescue Exception => error
      puts "Ugh oh. Something went wrong: #{error.message}"
    end
  end
  
  nil
end

## Examples

minilang('PRINT')
# 0

minilang('5 PUSH 3 MULT PRINT')
# 15

minilang('5 PRINT PUSH 3 PRINT ADD PRINT')
# 5
# 3
# 8

minilang('5 PUSH POP PRINT')
# 5

minilang('3 PUSH 4 PUSH 5 PUSH PRINT ADD PRINT POP PRINT ADD PRINT')
# 5
# 10
# 4
# 7

minilang('3 PUSH PUSH 7 DIV MULT PRINT ')
# 6

minilang('4 PUSH PUSH 7 MOD MULT PRINT ')
# 12

minilang('-3 PUSH 5 SUB PRINT')
# 8

minilang('6 PUSH')
# (nothing printed; no PRINT commands)

minilang('FNORD')

# minilang('POP DIV')

File: 210/small-problems/medium1/stack-machine.js

function execute(program) {
  var stack = [];
  var register = 0;

  var checkStackIsNotEmpty = function () {
    if (stack.length === 0) {
      throw new Error('Stack is empty!');
    }
  }

  program.split(' ').forEach(function (instruction) {
    if (!!Number(instruction)) {
      register = Number(instruction);
      return;
    }

    switch (instruction) {
      case 'PRINT': {
        console.log(register);
        break;
      }
      case 'PUSH': {
        stack.push(register);
        break;
      }
      case 'POP': {
        checkStackIsNotEmpty();
        register = stack.pop();
        break;
      }
      case 'ADD': {
        checkStackIsNotEmpty();
        register += stack.pop();
        break;
      }
      case 'SUB': {
        checkStackIsNotEmpty();
        register -= stack.pop();
        break;
      }
      case 'MULT': {
        checkStackIsNotEmpty();
        register *= stack.pop();
        break;
      }
      case 'DIV': {
        checkStackIsNotEmpty();
        register = Math.floor(register / stack.pop());
        break;
      }
      case 'MOD': {
        checkStackIsNotEmpty();
        register = Math.floor(register % stack.pop());
        break;
      }
      default: throw new SyntaxError('Invalid instruction: "' + instruction + '"');
    }
  });
}

// Examples

(execute('PRINT'));
// 0

(execute('5 PUSH 3 MULT PRINT'));
// 15

(execute('5 PRINT PUSH 3 PRINT ADD PRINT'));
// 5
// 3
// 8

(execute('5 PUSH POP PRINT'));
// 5

(execute('3 PUSH 4 PUSH 5 PUSH PRINT ADD PRINT POP PRINT ADD PRINT'));
// 5
// 10
// 4
// 7

(execute('3 PUSH PUSH 7 DIV MULT PRINT'));
// 6

(execute('4 PUSH PUSH 7 MOD MULT PRINT'));
// 12

(execute('-3 PUSH 5 SUB PRINT'));
// 8

(execute('6 PUSH'));
//
// (nothing is printed because the `program` argument has no `PRINT` commands)
